Bti doc (#152)
OP-TEE/optee_docs: architecture/arm_security_extensions.rst

.. _arm_security_extensions:

#######################
Arm Security Extensions
#######################

.. _bti:

Branch Target Identification
****************************

Branch Target Identification (BTI) is an ARMv8.5 extension that provides
Control Flow Integrity (CFI) around indirect branches and their targets, thus helping
to limit the JOP (Jump Oriented Programming) attacks.

With this extension, ARM8.5-A introduces Branch Target Instructions (BTIs).

> BTIs are also called launching pads. The processor can be configured so that

Reviewer: s/launching/landing/, no?

Author: Fixed..Thanks

indirect branches (BR and BLR) only allows target landing pad instructions.
If the target of an indirect branch is not a landing pad, a Branch Target Exception 
is generated.

How to enable BTI for OP-TEE core
==================================

To make use of BTI in TEE core on CPU's that support it, enable the option 

> ``CFG_CORE_BTI.``

Reviewer 2: ```suggestion
``CFG_CORE_BTI``.
```

OP-TEE core makes use of some built-ins in the GCC/clang toolchains. So, in order
to use the option ``CFG_CORE_BTI``, make sure that GCC toolchain has been built with 
``--enable-standard-branch-protection`` is used else OP-TEE will fail to build.

> By default libraries such as libgcc.a are built with default flags (``-mbranch-protection=none``),

Reviewer 3: _By default ... with default flags._
I guess one of the defaults can be removed.

hence are incompatible with branch protection enabled. The Arm GNU compiler team
is looking for ways of providing users easy access to BTI-enabled libraries. 
In the short-term, they plan to create documentation to make it easier for users to
build BTI-enabled libraries themselves. Longer-term, they will begin discussions
on how to ensure BTI-enabled libraries are available automatically to users.
Please contact GCC team for more information on same.

The same problem is also there with clang toolchain. So, when using clang to build

> OP-TEE with ``CFG_CORE_BTI=y``, builtins(found in llvm's "compiler-rt"

Reviewer 2: ```suggestion
OP-TEE with ``CFG_CORE_BTI=y``, builtins (found in llvm's "compiler-rt"
```

project) must be built with BTI protection enabled. We have some instructions on
how to build the compiler-rt with BTI enabled. These are available in FAQ.

How to enable BTI for TA's
===========================

To make use of BTI support for TA's and user mode libraries, enable the option
``CFG_TA_BTI``. This will ensure that all libraries provided by OP-TEE to the TA's
as well as the TA's are built with BTI option.

When the TA's are loaded by ldelf, they are checked at run time for the BTI NOTE
property in ELF before enabling the protection for the TA.

> When building TA's, you need to ensure that the any external library used has been

Reviewer 3: _ensure that the any external_
Please drop "the".

.. _arm_security_extensions:

#######################

> ARM Security Extensions

Reviewer 2: ```suggestion
Arm Security Extensions
```

#######################

.. _bti:

Branch Target Identification
****************************

Branch Target Identification (BTI) is an ARMv8.5 extension that provides

> control flow integrity around indirect branches and their targets, thus helping

Reviewer 2: ```suggestion
Control Flow Integrity (CFI) around indirect branches and their targets, thus helping
```

> to limit the JOP (Jump oriented programming) attacks.

Reviewer 2: ```suggestion
to limit the JOP (Jump Oriented Programming) attacks.
```

> With this extension, ARM8.5-A introduces Brach Target Instructions (BTIs).

Reviewer 2: ```suggestion
With this extension, ARM8.5-A introduces Branch Target Instructions (BTIs).
```

> BTIs are also called launching padis. The processor can be configured so that

Reviewer 2: ```suggestion
BTIs are also called launching pads. The processor can be configured so that
```

> indirect branches (BR and BLR) can only allow target landing pad instructions.

Reviewer 2: ```suggestion
indirect branches (BR and BLR) only allows target landing pad instructions.
```

If the target of an indirect branch is not a landing pad, a Branch Target Exception 
is generated.

How to enable BTI for OP-TEE core
==================================

> To make use of BTI in TEE cores on CPU's that support it, enable the option 

Reviewer 2: ```suggestion
To make use of BTI in TEE core on CPU's that support it, enable the option 
```

> CFG_CORE_BTI.

Reviewer 2: ```suggestion
``CFG_CORE_BTI.``
```

> OP-TEE core makes use of some built-ins in the GCC/clag toolchains. So, inorder

Reviewer 2: ```suggestion
OP-TEE core makes use of some built-ins in the GCC/clang toolchains. So, in order
```

> to use the option CFG_CORE_BTI, make sure that GCC toolchain built with 

Reviewer 2: ```suggestion
to use the option CFG_CORE_BTI, make sure that GCC toolchain has been built with 
```

> to use the option CFG_CORE_BTI, make sure that GCC toolchain built with 

Reviewer 2: ```suggestion
to use the option ``CFG_CORE_BTI``, make sure that GCC toolchain built with 
```

> --enable-standard-branch-protection is used else op-tee will fail to build.

Reviewer 2: ```suggestion
``--enable-standard-branch-protection`` is used else OP-TEE will fail to build.
```

> By default libraries such as libgcc.a are built with default flags (-mbranch-protection=none),

Reviewer 2: ```suggestion
By default libraries such as libgcc.a are built with default flags (``-mbranch-protection=none``),
```

hence are incompatible with branch protection enabled. The Arm GNU compiler team
is looking for ways of providing users easy access to BTI-enabled libraries. 
In the short-term, they plan to create documentation to make it easier for users to
build BTI-enabled libraries themselves. Longer-term, they will begin discussions
on how to ensure BTI-enabled libraries are available automatically to users.
Please contact GCC team for more information on same.

The same problem is also there with clang toolchain. So, when using clang to build

> OP-TEE with CFG_CORE_BTI=y, builtins(found in llvm's "compiler-rt"

Reviewer 2: ```suggestion
OP-TEE with ``CFG_CORE_BTI=y``, builtins(found in llvm's "compiler-rt"
```

> project) must be built with bti protection enabled. We have some instructions on

Reviewer 2: ```suggestion
project) must be built with BTI protection enabled. We have some instructions on
```

how to build the compiler-rt with BTI enabled. These are available in FAQ.

How to enable BTI for TA's
===========================

To make use of BTI support for TA's and user mode libraries, enable the option

> CFG_TA_BTI. This will ensure that all libraries provided by op-tee to the TA's

Reviewer 2: ```suggestion
``CFG_TA_BTI``. This will ensure that all libraries provided by OP-TEE to the TA's
```

as well as the TA's are built with BTI option.

When the TA's are loaded by ldelf, they are checked at run time for the BTI NOTE
property in ELF before enabling the protection for the TA.

> When building TA, you need to ensure that the any external library used has been

Reviewer 2: ```suggestion
When building TA's, you need to ensure that the any external library used has been
```

built with branch-protection. This can be done by checking the library using readelf

> command with option -n. The BTI enabled libraries will have BTI NOTE property in

Reviewer 2: ```suggestion
command with option ``-n``. The BTI enabled libraries will have BTI NOTE property in
```

> .note.gnu.property section. If that is not the case, compilation will stop with a

Reviewer 2: ```suggestion
``.note.gnu.property`` section. If that is not the case, compilation will stop with a
```

warning. This is done intentionally to warn the user.

> Note

Reviewer 2: Note (pardon the pun), that you can write rst-notes like [this](https://github.com/OP-TEE/optee_docs/blame/master/building/gits/optee_os.rst#L98-L101) that in turn will render like [this](https://optee.readthedocs.io/en/latest/building/gits/optee_os.html#build-using-llvm-clang).

====

> The BTI support is currently not compatible with options CFG_VIRTUALIZATION and

Reviewer 2: ```suggestion
The BTI support is currently not compatible with options ``CFG_VIRTUALIZATION`` and
```

> CFG_WITH_PAGER.

Reviewer 2: ```suggestion
``CFG_WITH_PAGER``.
```
